src/ctl/ctl.c

```diff
@@ -0,0 +1,579 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/* Copyright 2016-2024, Intel Corporation */
+
+/*
+ * ctl.c -- implementation of the interface for examination and modification of
+ *	the library's internal state
+ */
+#include "ctl.h"
+#include "alloc.h"
+#include "os.h"
+#include <ctype.h>
+#include <limits.h>
+#include <stdint.h>
+#include <string.h>
+// #include "core_assert.h"
+
+#define CTL_MAX_ENTRIES 100
+
+#define MAX_CONFIG_FILE_LEN (1 << 20) /* 1 megabyte */
+
+#define CTL_STRING_QUERY_SEPARATOR ";:"
+#define CTL_NAME_VALUE_SEPARATOR "="
+#define CTL_QUERY_NODE_SEPARATOR "."
+#define CTL_VALUE_ARG_SEPARATOR ","
+
+static int ctl_global_first_free = 0;
+static struct ctl_node CTL_NODE(global)[CTL_MAX_ENTRIES];
+
+/*
+ * This is the top level node of the ctl tree structure. Each node can contain
+ * children and leaf nodes.
+ *
+ * Internal nodes simply create a new path in the tree whereas child nodes are
+ * the ones providing the read/write functionality by the means of callbacks.
+ *
+ * Each tree node must be NULL-terminated, CTL_NODE_END macro is provided for
+ * convenience.
+ */
+struct ctl {
+    struct ctl_node root[CTL_MAX_ENTRIES];
+    int first_free;
+};
+
+/*
+ * ctl_find_node -- (internal) searches for a matching entry point in the
+ *	provided nodes
+ *
+ * The caller is responsible for freeing all of the allocated indexes,
+ * regardless of the return value.
+ */
+static const struct ctl_node *ctl_find_node(const struct ctl_node *nodes,
+                                            const char *name,
+                                            struct ctl_indexes *indexes) {
+    // LOG(3, "nodes %p name %s indexes %p", nodes, name, indexes);
+
+    const struct ctl_node *n = NULL;
+    char *sptr = NULL;
+    char *parse_str = Strdup(name);
+    if (parse_str == NULL) {
+        return NULL;
+    }
+
+    char *node_name = strtok_r(parse_str, CTL_QUERY_NODE_SEPARATOR, &sptr);
+
+    /*
+	 * Go through the string and separate tokens that correspond to nodes
+	 * in the main ctl tree.
+	 */
+    while (node_name != NULL) {
+        char *endptr;
+        /*
+		 * Ignore ERANGE or EINVA errno from strtol.
+		 * endptr is used to check conversion result.
+		 */
+        int tmp_errno = errno;
+        long index_value = strtol(node_name, &endptr, 0);
+        errno = tmp_errno;
+        struct ctl_index *index_entry = NULL;
+        if (endptr != node_name) { /* a valid index */
+            index_entry = Malloc(sizeof(*index_entry));
+            if (index_entry == NULL) {
+                goto error;
+            }
+            index_entry->value = index_value;
+            PMDK_SLIST_INSERT_HEAD(indexes, index_entry, entry);
```

Comment:
PMDK? 😉

Reply:
Will be removed in the future, as part of replacing list implementation.

Reply:
Done